src/ops/convolution.rs

```diff
@@ -0,0 +1,495 @@
+            input_dilation.len(),
+            groups,
+            flip,
+            stream.as_ptr(),
+        ))
+    }
+}
+
+/// General convolution over an input with several channels.
+///
+/// > Only 1d and 2d convolutions are supported at the moment
+///
+/// > the default `groups: 1` is currently supported
+///
+/// # Params
+/// - array: Input array of shape `&[N, ..., C_in]`
+/// - weight: Weight array of shape `&[C_out, ..., C_in]`
+/// - strides: The kernel strides. All dimensions get the same stride if only one number is specified.
+/// - padding: The input padding. All dimensions get the same padding if only one number is specified.
+/// - kernel_dilation: `The  kernel dilation. All dimensions get the same dilation if only one number is specified.
+/// - input_dilation: The input dilation. All dimensions get the same dilation if only one number is specified.
+/// - groups: Input feature groups
+/// - flip: Flip the order in which the spatial dimensions of the weights are processed.
+///   Performs the cross-correlation operator when `flip` is `false` and the convolution
+///   operator otherwise.
+/// - stream: stream or device to evaluate on
+#[default_device]
+pub fn conv_general_device<'a>(
+    array: &Array,
+    weight: &Array,
+    strides: impl Into<Option<&'a [i32]>>,
+    padding: impl Into<Option<&'a [i32]>>,
+    kernel_dilation: impl Into<Option<&'a [i32]>>,
+    input_dilation: impl Into<Option<&'a [i32]>>,
+    groups: impl Into<Option<i32>>,
+    flip: impl Into<Option<bool>>,
+    stream: StreamOrDevice,
+) -> Array {
+    try_conv_general_device(
+        array,
+        weight,
+        strides,
+        padding,
+        kernel_dilation,
+        input_dilation,
+        groups,
+        flip,
+        stream,
+    )
+    .unwrap()
+}
+
+/// General convolution over an input with several channels without validating the inputs.
+///
+/// > Only 1d and 2d convolutions are supported at the moment
+///
+/// > the default `groups: 1` is currently supported
+///
+/// # Params
+/// - array: Input array of shape `&[N, ..., C_in]`
+/// - weight: Weight array of shape `&[C_out, ..., C_in]`
+/// - strides: The kernel strides. All dimensions get the same stride if only one number is specified.
+/// - padding: The input padding. All dimensions get the same padding if only one number is specified.
+/// - kernel_dilation: `The  kernel dilation. All dimensions get the same dilation if only one number is specified.
+/// - input_dilation: The input dilation. All dimensions get the same dilation if only one number is specified.
+/// - groups: Input feature groups
+/// - flip: Flip the order in which the spatial dimensions of the weights are processed.
+///   Performs the cross-correlation operator when `flip` is `false` and the convolution
+///   operator otherwise.
+/// - stream: stream or device to evaluate on
+///
+/// # Safety
+///
+/// This function is unsafe because it does not validate the inputs.
+///
+///  # Panic
```

Comment:
Does it panic in a way that rust runtime can catch or is it a fatal error that cannot be caught by the rust runtime?

Reply:
Updated to denote what i meant, no #panic header